socorro/external/postgresql/products.py

```diff
@@ -14,11 +18,131 @@
+            operator_ = value[:2]
+            value = value[2:]
+        elif any(itertools.imap(value.startswith, ('=', '>', '<'))):
+            operator_ = value[:1]
+            value = value[1:]
+        else:
+            operator_ = '='
+        return (operator_, string_to_datetime(value).date())
+
+
+class ProductVersions(PostgreSQLBase):
+
+    def get(self, **kwargs):
+        filters = [
+            ("version", None, [str]),
+            ("product", None, [str]),
+            ("featured", None, bool),
+            ("start_date", None, SmartDate()),
+            ("end_date", None, SmartDate()),
+            ("active", None, bool),
+            ("is_rapid_beta", None, bool),
+        ]
+        params = external_common.parse_arguments(filters, kwargs, modern=True)
+        where = []
+        sql_params = {}
+        for param, value in params.items():
+            if value is None:
+                continue
+            param = {
+                'product': 'product_name',
+                'version': 'version_string',
+                'featured': 'is_featured',
+            }.get(param, param)
+
+            if param == 'active':
+                # This is a convenient exception. It makes it possible
+                # to query for only productversions that are NOT sunset
+                # without having to do any particular date arithmetic.
+                param = 'end_date'
+                operator_ = value and '>' or '<='
+                value = datetime.datetime.utcnow().date().isoformat()
+            elif isinstance(value, list):
+                operator_ = 'IN'
+                value = tuple(value)
+            elif isinstance(value, tuple):
+                assert len(value) == 2
+                operator_, value = value
+            else:
+                operator_ = '='
+            where.append('{} {} %({})s'.format(
+                param,
+                operator_,
+                param
+            ))
+            sql_params[param] = value
+
+        # rewrite it to a string
+        if where:
+            sql_where = 'WHERE ' + ' AND '.join(where) + '\n'
+        else:
+            sql_where = ''
+
+        # XXX TODO
+        # RE-write this to use the same SQL select that's used by the
+        # the `product_info` view.
+        sql = """
+            /* socorro.external.postgresql.products.ProductVersions.get */
+            SELECT
+                product_name AS product,
+                version_string AS version,
+                start_date,
+                end_date,
+                throttle::REAL,
+                is_featured AS featured,
+                build_type,
+                has_builds,
+                is_rapid_beta
+            FROM product_info
+            {}
+            ORDER BY product_sort, version_sort DESC, channel_sort
+        """.format(sql_where)
+        # print sql
+        # print sql_params
+        results = self.query(sql, sql_params).zipped()
```

Comment:
``self.query`` has an ``error_message`` argument that is used to show better errors in case of failure.